### PR TITLE
arch/arm64/imx9: Add system reset controller

### PR DESCRIPTION
**Pull request by jlaitine**
System reset conttoller to powercycle ml and media blocks and disable power-isolation

## Summary

Adds supporting functions for system reset controller

## Impact

Only for imx9

## Testing

Functions tested on a custom i.MX93 board

### CODE REVIEW

**Review by acassis — APPROVED** [review]

File: arch/arm64/src/imx9/imx9_system_ctl.c [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+ ****************************************************************************/
+
+void imx9_mix_powerup(void)
+{
+  uint32_t val = 0;
+
+  /* Authen ctrl, enable NS access to slice registers */
+
+  modifyreg32(IMX9_SRC_MEDIA_SLICE_BASE
+    + SRC_SLICE_AUTHEN_CTRL_OFFSET, 0, BIT(9));
```
Comment (acassis) [inline arch/arm64/src/imx9/imx9_system_ctl.c:50]: Please align with previous line, the "+" should be just after "("

File: arch/arm64/src/imx9/imx9_system_ctl.c [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+  modifyreg32(IMX9_SRC_ML_SLICE_BASE
+    + SRC_SLICE_AUTHEN_CTRL_OFFSET, 0, BIT(9));
```
Comment (acassis) [inline arch/arm64/src/imx9/imx9_system_ctl.c:52]: Ditto

File: arch/arm64/src/imx9/imx9_system_ctl.c [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+  /* Enable s400 handsake */
+
+  modifyreg32(IMX9_BLK_CTRL_S_AONMIX2_BASE
+    + AON_MIX_LP_HANDSAKE, 0, BIT(13));
```
Comment (acassis) [inline arch/arm64/src/imx9/imx9_system_ctl.c:57]: ditto

File: arch/arm64/src/imx9/imx9_system_ctl.c [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+  /* Counter mode */
+
+  modifyreg32(IMX9_SRC_MEDIA_SLICE_BASE
+    + SRC_SLICE_PSW_ACK_CTRL0_OFFSET, BIT(28), BIT(29));
```
Comment (acassis) [inline arch/arm64/src/imx9/imx9_system_ctl.c:62]: ditto

File: arch/arm64/src/imx9/imx9_system_ctl.c [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+  modifyreg32(IMX9_SRC_ML_SLICE_BASE
+    + SRC_SLICE_PSW_ACK_CTRL0_OFFSET, BIT(28), BIT(29));
```
Comment (acassis) [inline arch/arm64/src/imx9/imx9_system_ctl.c:64]: ditto

File: arch/arm64/src/imx9/imx9_system_ctl.c [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+  /* release media and ml from reset */
+
+  modifyreg32(IMX9_SRC_GENERAL_REG_BASE
+    + SRC_CTRL_OFFSET, 0, (BIT(4) | BIT(5)));
```
Comment (acassis) [inline arch/arm64/src/imx9/imx9_system_ctl.c:69]: ditto

File: arch/arm64/src/imx9/imx9_system_ctl.c [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+  /* Enable mem in Low power auto sequence */
+
+  modifyreg32(IMX9_SRC_MEDIA_MEM_BASE + MEM_CTRL_OFFSET, 0, BIT(2));
+  modifyreg32(IMX9_SRC_ML_MEM_BASE + MEM_CTRL_OFFSET, 0, BIT(2));
+
+  /* Mediamix powerdown */
+
+  modifyreg32(IMX9_SRC_MEDIA_SLICE_BASE
+    + SRC_SLICE_SW_CTRL_OFFSET, 0, BIT(31));
```
Comment (acassis) [inline arch/arm64/src/imx9/imx9_system_ctl.c:79]: ditto

File: arch/arm64/src/imx9/imx9_system_ctl.c [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        val = getreg32(IMX9_SRC_MEDIA_SLICE_BASE
+          + SRC_SLICE_FUNC_STAT_OFFSET);
+
+      up_udelay(1);
+    }
+  else
+    {
+      while (!(val & BIT(0)))
+        val = getreg32(IMX9_SRC_MEDIA_SLICE_BASE
+          + SRC_SLICE_FUNC_STAT_OFFSET);
```
Comment (acassis) [inline arch/arm64/src/imx9/imx9_system_ctl.c:94]: ditto

File: arch/arm64/src/imx9/imx9_system_ctl.c [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+  modifyreg32(IMX9_SRC_MEDIA_SLICE_BASE
+    + SRC_SLICE_SW_CTRL_OFFSET, 0, BIT(31));
+
+  val = getreg32(IMX9_SRC_MEDIA_SLICE_BASE + SRC_SLICE_FUNC_STAT_OFFSET);
+  if (val & 1)
+    {
+      while (!(val & BIT(12)))
+        val = getreg32(IMX9_SRC_MEDIA_SLICE_BASE
+          + SRC_SLICE_FUNC_STAT_OFFSET);
```
Comment (acassis) [inline arch/arm64/src/imx9/imx9_system_ctl.c:86]: ditto

File: arch/arm64/src/imx9/imx9_system_ctl.c [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+    {
+      while (!(val & BIT(0)))
+        val = getreg32(IMX9_SRC_MEDIA_SLICE_BASE
+          + SRC_SLICE_FUNC_STAT_OFFSET);
+    }
+
+  /* Power on */
+
+  modifyreg32(IMX9_SRC_MEDIA_SLICE_BASE
+    + SRC_SLICE_SW_CTRL_OFFSET, BIT(31), 0);
```
Comment (acassis) [inline arch/arm64/src/imx9/imx9_system_ctl.c:100]: ditto

File: arch/arm64/src/imx9/imx9_system_ctl.c [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+  /* ML powerdown */
+
+  modifyreg32(IMX9_SRC_ML_SLICE_BASE + SRC_SLICE_SW_CTRL_OFFSET, 0, BIT(31));
+
+  val = getreg32(IMX9_SRC_ML_SLICE_BASE + SRC_SLICE_FUNC_STAT_OFFSET);
+  if (val & 1)
+    {
+      while (!(val & BIT(12)))
+        val = getreg32(IMX9_SRC_ML_SLICE_BASE + SRC_SLICE_FUNC_STAT_OFFSET);
```
Comment (acassis) [inline arch/arm64/src/imx9/imx9_system_ctl.c:112]: Please include "{" "}"

File: arch/arm64/src/imx9/imx9_system_ctl.c [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+      up_udelay(1);
+    }
+  else
+    {
+      while (!(val & BIT(0)))
+        val = getreg32(IMX9_SRC_ML_SLICE_BASE + SRC_SLICE_FUNC_STAT_OFFSET);
```
Comment (acassis) [inline arch/arm64/src/imx9/imx9_system_ctl.c:118]: Ditto

File: arch/arm64/src/imx9/imx9_system_ctl.c [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+  /* Power on */
+
+  modifyreg32(IMX9_SRC_ML_SLICE_BASE + SRC_SLICE_SW_CTRL_OFFSET, BIT(31), 0);
+  while (!(val & BIT(4)))
+      val = getreg32(IMX9_SRC_ML_SLICE_BASE + SRC_SLICE_FUNC_STAT_OFFSET);
```
Comment (acassis) [inline arch/arm64/src/imx9/imx9_system_ctl.c:125]: Ditto

File: arch/arm64/src/imx9/imx9_system_ctl.c [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+
+  /* Mediamix powerdown */
+
+  modifyreg32(IMX9_SRC_MEDIA_SLICE_BASE
+    + SRC_SLICE_SW_CTRL_OFFSET, 0, BIT(31));
+
+  val = getreg32(IMX9_SRC_MEDIA_SLICE_BASE + SRC_SLICE_FUNC_STAT_OFFSET);
+  if (val & 1)
+    {
+      while (!(val & BIT(12)))
```
Comment (acassis) [inline arch/arm64/src/imx9/imx9_system_ctl.c:84]: Please include "{" "}"

**Comment by jlaitine** [comment]
Thank you! The style should now be fixed.

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**